CIPs/0003.md

```diff
@@ -0,0 +1,134 @@
+# CIP [0003]: Identity Claims and Metadata
+
+- Date: 2019-09-19
+- Author: @timmoreton
+- Status: WIP
+
+## Overview
+
+This CIP describes a general mechanism that allows principals on the Celo platform to claim real-world identities and maintain rich, unverified metadata. Metadata may be public, or optionally, encrypted and accessible only to the Celo Foundation. Third parties can verify these identity claims, and use the associated unverified metadata as appropriate. 
+
+There are a number of principals that provide resources to or deploy services on the Celo platform that may wish to assert identities or provide metadata. These include, but are not limited to:
+* Validators
+* Validator groups
+* Smart contract owners
+* DApp developers
+* Creators of Governance proposals
+* Oracles 
+
+Use cases include:
+* Tools want to present public metadata supplied by a validator or validator group as part of a list of candidate groups, or a list of current elected validators.
+* Governance Explorer UIs may want to present public metadata about the creators of governance proposals
+* The Celo Foundation receives notice of a security vulnerability and wants to contact elected validators to facilitate them to make a decision on applying a patch.
+* A DApp makes a request to the Celo Wallet for account information or to sign a transaction. The Celo Wallet should provide information about the DApp to allow the user to make a decision whether to accept the transaction or not.   
+
+## Goals
+
+- Integrate with the existing Attestations structure, rather than maintain separate identifer fields in other contracts, so that the same identity claims and metadata can be linked from any of a principals' activities  
+- Allow principals to make an extensible range of identity claims, and support clients verifying the veracity of these 
+- Allow principals to supply unverified extensible, structured metadata off-chain
+- Allow tools to choose what metadata they require and how they present it
+- Support both public (unencrypted) metadata and metadata accessible only to the Celo Foundation
+
+## Proposed Solution
+
+### Claims
+
+The Attestations contract is extended as follows:
+*  Add a `struct Claim` that comprises: 
+   * `service` - Required. Initially supported values for `service` include `primary`, `github`, and `twitter`.
+   * `identifier` - Optional. The identifier being claimed on the service being claimed.  
+   * `data` - Optional. Additional data, e.g. a 
+   URL, to locate a linked signed reference to the claim. 
+* A field array `Claim[] claims` is added to the `Account` struct. 
+
+All claims are optional.
+
+### Primary Claims and Public Metadata
+
+For claims with `service` set to `primary`, `identifier` is a domain name. `data` is an optional path prefix, such that `https://<identifier>/<data>/identity.json` is a file in the following format:
```

Comment:
Related to the above comment, what was the reasoning to particularly denote a DNS claim as primary?

Reply:
Both Google and Keybase look like the prove ownership of a domain via TXT Record. 

I wonder if they do that because it might be technically more secure than http(s), as some websites might allow users to host data at controllable paths? A mitigation could be what Let's encrypt and DNT do, to use [RFC 5785](https://tools.ietf.org/html/rfc5785) with `domain/.well-known`. Fun fact stellar also uses that folder https://www.stellar.org/developers/guides/concepts/stellar-toml.html

Or because a naked top level domain might prescribe users more infrastructure demands with a webserver than simply editing a TXT record? On the other hand, making HTTP calls is probably easier than querying for TXT records, though there are now DNS over HTTPS servers, so it would actually preserve privacy a little better this way.

Reply:
Mostly as a personal anecdote, but with my personal setup at nambrot.com, it would kind of a pain to host it on my specific path. I'd rather prefer to be able to specify under which path (any domain) I would like to host my public metadata at.

It separates the concern of proving a claim from hosting data with the identity of that claim.